dash/_callback_context.py

```diff
@@ -152,7 +152,7 @@ def display(btn1, btn2):
         grouping = getattr(flask.g, "args_grouping", {})
 
         def update_args_grouping(g):
-            if isinstance(g, dict) and "id" in g:
+            if isinstance(g, dict) and "id" in g and "property" in g:
```

Comment:
What if someone makes a PMC ID with both `id` AND `property` as keys? Is there some other way we can tell the difference between a dict we should recurse into and a dict ID?

Reply:
D'oh!  
ok, I think it's correct in the 3rd update :slightly_smiling_face:

Reply:
I'm still worried that there could be ways to fool us - now perhaps we wouldn't recurse when we should. Like:
```py
@callback(
    inputs=dict(
        a=dict(id=Input(...), property=Input(...))
    )
)
```

Feels to me like the more robust way to handle this would be when we have a known flat list [`inputs_state`](https://github.com/plotly/dash/blob/c12c957d5fff8d0d65ca06276af7f9e90b424971/dash/dash.py#L1431-L1432) - add the extra info to THAT before we map that into its grouped form.

Separately, would it be worthwhile augmenting `outputs_grouping` as well? That doesn't have `triggered` or `value` but it could benefit from `str_id` and having all its parts be `AttributeDict`s

Reply:
Your above example works fine, so do all of the following examples. So far, I haven't found anything that _didn't_ work.

As far as `outputs_grouping`, I think it's a good idea, and I agree that it should be a separate issue.  



--------------
```python
from dash import Dash, html, Input, Output,ctx
app = Dash(__name__, suppress_callback_exceptions=True)

app.layout = html.Div([
    html.Button("btn1",id="btn1", n_clicks=0),
    html.Button("btn2", id="btn2", n_clicks=0),
    html.Div(id='container', children=[]),
])


@app.callback(
    Output('container', 'children'),
    inputs=dict(
        a=dict(id=Input("btn1", "n_clicks"), property=Input("btn2", "n_clicks")),
    )
)
def update1(a):
    g=ctx.args_grouping
    return  f"btn1 clicked {g.a.id.value} times, btn2 clicked {g.a.property.value} times"


if __name__ == '__main__':
    app.run_server(debug=True)

```

This example has component ids as "id" and "property" as well

-----------------
```python

from dash import Dash, html, Input, Output, ctx
app = Dash(__name__, suppress_callback_exceptions=True)

app.layout = html.Div([
    html.Button("id",id="id", n_clicks=0),
    html.Button("property", id="property", n_clicks=0),
    html.Div(id='container', children=[]),
])


@app.callback(
    Output('container', 'children'),
    inputs=dict(
        a=dict(id=Input("id", "n_clicks"), property=Input("property", "n_clicks")),
    )
)
def update1(a):
    g=ctx.args_grouping
    return  f"id btn clicked {g.a.id.value} times, property btn clicked {g.a.property.value} times"


if __name__ == '__main__':
    app.run_server(debug=True)


```

------

Pattern matching dict id's are fine too:

```python
from dash import Dash, html, Input, Output, ctx
app = Dash(__name__, suppress_callback_exceptions=True)

app.layout = html.Div([
    html.Button("id",id="id", n_clicks=0),
    html.Button("property", id="property", n_clicks=0),
    html.Div(id='container', children=[]),
])


@app.callback(
    Output('container', 'children'),
    inputs=dict(
        a=dict(id=Input("id", "n_clicks"), property=Input("property", "n_clicks")),
    )
)
def update1(a):
    g=ctx.args_grouping
    return  f"id btn clicked {g.a.id.value} times, property btn clicked {g.a.property.value} times"


if __name__ == '__main__':
    app.run_server(debug=True)
```